File: Python/Greedy/22_smallest_subset_with_greater_sum.py
"""
Idea:
sorting in descending order,  finding the sum of total array.
Then, from start, sum till difference between temporary sum and original sum is greater than temporary sum.
"""


# User function Template for python3

class Solution:

    # Function to find the minimum subset with the given conditions.
    def minSubset(self, A, N):

        # Calculating the sum of all elements in the list.
        s = sum(A)
        res = 0

        # Sorting the list in descending order.
        A.sort()
        A.reverse()

        # Iterating through the list.
        for i in range(0, N):

            # Adding the current element to the result subset.
            res += A[i]

            # Subtracting the current element from the total sum.
            s -= A[i]

            # Checking if the sum of the result subset is greater than the remaining sum. If yes, returning the number of elements in the result subset.
            if res > s:
                return i + 1

        # Returning the total number of elements in the list if no subset is found.
        return N


# {
# Driver Code Starts
# Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        N = int(input())
        A = list(map(int, input().strip().split()))
        ob = Solution()
        ans = ob.minSubset(A, N)
        print(ans)
# } Driver Code Ends

File: C++/Greedy/22_smallest_subset_with_greater_sum.cpp
/*
Idea:
sorting in descending order,  finding the sum of total array.
Then, from start, sum till difference between temporary sum and original sum is greater than temporary sum.
*/

//{ Driver Code Starts
//Initial Template for C++

#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function Template for C++

class Solution {
public:
    //Function to find the minimum number of elements in the first subset.
    int minSubset(vector<int>& Arr, int N) {
        long long sum = 0;

        //calculating the sum of all elements in the Arr vector.
        for (int i = 0;i < N;i++) {
            sum += Arr[i];
        }

        //sorting the Arr vector in descending order.
        sort(Arr.rbegin(), Arr.rend());
        long long res = 0;

        //iterating through the Arr vector to find the minimum number of elements in the first subset.
        for (int i = 0;i < N;i++) {
            res += Arr[i];
            sum -= Arr[i];

            //if the sum of elements in the first subset becomes greater than the sum of remaining elements,
            //return the index i+1 which represents the minimum number of elements in the first subset.
            if (res > sum) {
                return i + 1;
            }
        }
        //if no such index is found, return N which represents that all elements are in the first subset.
        return N;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        int N;
        cin >> N;
        vector<int> Arr(N);
        for (int i = 0;i < N;i++) {
            cin >> Arr[i];
        }
        Solution ob;
        cout << ob.minSubset(Arr, N) << endl;
    }
    return 0;
}
// } Driver Code Ends